hooks/environment

```diff
@@ -1,107 +1,20 @@
 #!/bin/bash
-
-set -eu -o pipefail
+set -e -o pipefail -u
 
 basedir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && cd .. && pwd )"
+credhelper="$basedir/git-credential-s3-secrets"
 
-# shellcheck disable=SC1090
-. "$basedir/lib/shared.bash"
-
-# For resiliency, increase the number of attempts to retrieve credentials for IAM roles
-# The default value is 1
-export AWS_METADATA_SERVICE_NUM_ATTEMPTS=3
```

Comment:
Good question. I imagine the Go SDK would have the same default retries as the CLI, and would probably respect the same environment to configure that. But it might also have a programatic way to change that. I had always wondered if this `export` would be leaking out of this script as one of the env changes for the whole build. Doing it within Go would better scope it.

Reply:
Hmm… https://github.com/aws/aws-sdk-go/issues/3495 … and no mention of Go in https://docs.aws.amazon.com/credref/latest/refdocs/setting-global-metadata_service_num_attempts.html

Reply:
I think I'm happy to merge anyway, because I think this patch fundamentally improves the situation instead of needing to patch over it with retries:

> [AWS_METADATA_SERVICE_NUM_ATTEMPTS] was recommended by aws-support for boto3/aws-cli due to IMDS being a service with rate limits.
— https://github.com/aws/aws-sdk-go/issues/3495#issuecomment-678256013

The bash version of this hook was hitting IMDS for credentials 16 times. The go version hits IMDS once. So I suspect we're 16x less likely to hit rate limit, and 16x less likely to be affected by a transient IMDS failure.

> **Before:**
> shell out to aws s3
> **request session credentials**
> HEAD bucket
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/private_ssh_key
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/private_ssh_key
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/id_rsa_github
> shell out to aws s3
> **request session credentials**
> GET pipeline/id_rsa_github
> start ssh-agent
> ssh-add
> shell out to aws s3
> **request session credentials**
> HEAD private_ssh_key
> shell out to aws s3
> **request session credentials**
> HEAD id_rsa_github
> shell out to aws s3
> **request session credentials**
> HEAD env
> shell out to aws s3
> **request session credentials**
> HEAD environment
> shell out to aws s3
> **request session credentials**
> GET environment
> eval new environment
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/env
> shell out to aws s3
> **request session credentials**
> GET pipeline/env
> eval new environment
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/environment
> shell out to aws s3
> **request session credentials**
> HEAD git-credentials
> shell out to aws s3
> **request session credentials**
> GET git-credentials
> shell out to aws s3
> **request session credentials**
> HEAD pipeline/git-credentials
> eval git-credentials environment
> 
> **After:**
> start go binary
> **request session credentials**
> HEAD bucket
> GET (concurrent) pipeline/private_ssh_key, pipeline/id_rsa_github, private_ssh_key, id_rsa_github, env, environment, pipeline/env, pipeline/environment, git-credentials, pipeline/git-credentials
> start ssh-agent
> ssh-add
> pass new environment (including ssh-agent and git-credentials env) back to bash script via stdout
> eval new environment